gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/DefaultTokenStateService.java

```diff
@@ -237,12 +249,37 @@ protected boolean isUnknown(final String token) {
 
   protected void updateExpiration(final String tokenId, long expiration) {
     synchronized (tokenExpirations) {
-      tokenExpirations.replace(tokenId, expiration);
+      if (tokenExpirations.containsKey(tokenId)) {
+        tokenExpirations.replace(tokenId, expiration);
+      } else {
+        tokenExpirations.put(tokenId, expiration);
+      }
     }
   }
 
   protected void removeToken(final String tokenId) throws UnknownTokenException {
     validateToken(tokenId);
+    removeTokenState(tokenId);
+  }
+
+  /**
+   * Bulk removal of the specified tokens.
+   *
+   * @param tokenIds The unique identifiers of the tokens whose state should be removed.
+   *
+   * @throws UnknownTokenException
+   */
+  protected void removeTokens(final Set<String> tokenIds) throws UnknownTokenException {
+    // Duplicating the logic from removeToken(String) here because this method is supposed to be an optimization for
+    // sub-classes that access an external store, for which bulk token removal performs better than individual removal.
+    // Sub-classes will have implemented removeToken(String), and calling that here will undo any optimizations provided
+    // by the sub-class's implementation of this method.
+    for (String tokenId : tokenIds) {
+      removeTokenState(tokenId);
+    }
+  }
+
+  private void removeTokenState(final String tokenId) {
```

Comment:
I'm not sure if I agree that any non-optimized code should be left if we already know that code piece will result in causing performance issues.
My two cents here:

1. in `removeToken` I'd call `removeTokens(Collections.singleton(tokenId))`
2. in `removeTokens` I'd **not** call a synchronized block `N` times. Instead, you may synchronize once, fetch the keyset and call the `removeAll` on that key set -> that collection is 

> backed by the map, so changes to the map are reflected in the set, and vice-versa.

Something like this:
```
    synchronized (tokenExpirations) {
      tokenExpirations.keySet().removeAll(tokenIds);
    }
```

Reply:
+1